### PR TITLE
Add skboutput function

### PR DESCRIPTION
**Pull request by xh4n3**
The skboutput idea was originally raised by @olsajiri in #1272 

Due to https://github.com/iovisor/bpftrace/pull/1272#issuecomment-1129282916, I rebased the code, improved a little bit. If we set BPFTRACE_PERF_RB_PAGES large enough, we now can capture as many packets as tcpdump does.

IMHO this may become increasingly useful as more tracepoints being added to the kernel, and the pcap file is much easier to dig into than any other method.

Please leave your comments below.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by olsajiri** [comment]
@xh4n3 please merge your changes from your last patch into previous patches as appropriate,
so we have clean changes without additional fixes

**Comment by xh4n3** [comment]
> @xh4n3 please merge your changes from your last patch into previous patches as appropriate, so we have clean changes without additional fixes

Thanks! I've rebased the commits.

File: man/adoc/bpftrace.adoc [hunk 1]
```diff
@@ -1756,6 +1756,47 @@ This function can only be used by functions that are allowed to, these functions
...
+=== skboutput
+
+.variants
+* `uint32 skboutput(const string path, struct sk_buff *skb, uint64 length, const uint64 offset)`
+
+*Kernel* 5.5
+
+*Helper* bpf_skb_output
+
+Write sk_buff `skb` data to a PCAP file in the `path`, starting from `offset` to `offset` + `length`.
```
Comment (fbs) [inline man/adoc/bpftrace.adoc:1768]: is this offset in the pcap data file or offset in the struct?

File: man/adoc/bpftrace.adoc [hunk 2]
```diff
@@ -1756,6 +1756,47 @@ This function can only be used by functions that are allowed to, these functions
...
+
```
Comment (fbs) [inline man/adoc/bpftrace.adoc:1769]: would be good to add how the pcap timestamp is determined

**Comment by fbs** [comment]
Thanks for picking this up, it would be a great feature to have.

Would it be possible for you to add tests for this feature too?

File: man/adoc/bpftrace.adoc (hunk 1 above)
Comment (xh4n3) [reply man/adoc/bpftrace.adoc:1768]: When bpf_skb_output is called on `dev_queue_xmit`'s context, the `skb`'s data section is prefixed by 14 bytes of ethernet header. On the context of `napi_gro_receive`, there is no ethernet header, only IP and upper layer headers.

So I added the `offset` arg to indicate how much bytes should we ignore from the beginning of skb's data, to make a legit Raw IP format pcap file. 

Is it okay to have this offset arg?  @fbs 

I'll updated the doc to make it clear.

**Comment by xh4n3** [comment]
Hi @fbs , I've updated the doc and added the tests. Would you please have another look? Thanks

**Review by fbs — APPROVED** [review]
looks good :)

File: src/bpffeature.cpp [hunk 3]
```diff
@@ -391,6 +391,61 @@ bool BPFfeature::has_kprobe_multi()
...
+  return bpf_create_map(
+      type, name.c_str(), key_size, value_size, max_entries, flags);
+#endif
+}
+
+bool BPFfeature::has_skb_output(void)
+{
+#ifndef HAVE_BCC_KFUNC
+  return false;
+#else
```
Comment (viktormalik) [inline src/bpffeature.cpp:414]: We don't have `HAVE_BCC_KFUNC` anymore, use `has_kfunc()`.

File: src/bpffeature.cpp [hunk 4]
```diff
@@ -391,6 +391,61 @@ bool BPFfeature::has_kprobe_multi()
...
+static int create_map(std::string name,
+                      enum bpf_map_type type,
+                      uint32_t key_size = sizeof(int),
+                      uint32_t value_size = sizeof(int),
+                      uint32_t max_entries = 1,
+                      uint32_t flags = 0)
+{
+#ifdef HAVE_BCC_CREATE_MAP
+  return bcc_create_map(
+      type, name.c_str(), key_size, value_size, max_entries, flags);
```
Comment (viktormalik) [inline src/bpffeature.cpp:403]: `bcc_create_map` shouldn't be used anymore. Check `detect_map` to see how map creation should be implemented.

**Review by viktormalik — COMMENTED** [review]
Sorry for a late review, overall LGTM.

The problem is that after merging #2265, there are some problems in `bpffeature.cpp` which must be fixed. I tried to put hits to appropriate places in the code.

File: src/ast/passes/codegen_llvm.cpp [hunk 5]
```diff
@@ -1114,6 +1114,36 @@ void CodegenLLVM::visit(Call &call)
...
+  else if (call.func == "skboutput")
+  {
+    auto elements = AsyncEvent::SkbOutput().asLLVMType(b_);
+    StructType *hdr_t = b_.GetStructType("hdr_t", elements, false);
+    AllocaInst *data = b_.CreateAllocaBPF(hdr_t, "hdr");
+
+    // The extra 0 here ensures the type of addr_offset will be int64
+    Value *aid_addr = b_.CreateGEP(data, { b_.getInt64(0), b_.getInt32(0) });
+    Value *id_addr = b_.CreateGEP(data, { b_.getInt64(0), b_.getInt32(1) });
+    Value *time_addr = b_.CreateGEP(data, { b_.getInt64(0), b_.getInt32(2) });
```
Comment (viktormalik) [inline src/ast/passes/codegen_llvm.cpp:1126]: These won't build on LLVM 14. See #2125 on ideas how to fix this.

**Review by viktormalik — CHANGES_REQUESTED** [review]

**Review by viktormalik — APPROVED** [review]

**Pull request merged**